nova_vm/src/ecmascript/execution/environments/global_environment.rs

```diff
@@ -1,68 +1,83 @@
+use std::collections::HashMap;
 
-#[derive(Debug, Clone, Copy)]
-pub struct GlobalEnvironmentIndex(NonZeroU32, PhantomData<GlobalEnvironment>);
-
-impl GlobalEnvironmentIndex {
-    pub const fn from_u32_index(value: u32) -> Self {
-        assert!(value != u32::MAX);
-        // SAFETY: Number is not max value and will not overflow to zero.
-        // This check is done manually to allow const context.
-        Self(unsafe { NonZeroU32::new_unchecked(value + 1) }, PhantomData)
-    }
-
-    pub const fn from_usize_index(value: usize) -> Self {
-        debug_assert!(value < u32::MAX as usize);
-        Self(
-            unsafe { NonZeroU32::new_unchecked(value as u32 + 1) },
-            PhantomData,
-        )
-    }
-
-    pub const fn into_index(self) -> usize {
-        self.0.get() as usize - 1
-    }
-}
+use super::{DeclarativeEnvironment, ObjectEnvironment};
 
-/// 9.1.1.4 Global Environment Records
-/// https://tc39.es/ecma262/#sec-global-environment-records
+/// ### [9.1.1.4 Global Environment Records](https://tc39.es/ecma262/#sec-global-environment-records)
+///
+/// A Global Environment Record is used to represent the outer most scope that
+/// is shared by all of the ECMAScript Script elements that are processed in a
+/// common realm. A Global Environment Record provides the bindings for built-in
+/// globals (clause 19), properties of the global object, and for all top-level
+/// declarations (8.2.9, 8.2.11) that occur within a Script.
 #[derive(Debug)]
 pub struct GlobalEnvironment {
-    /// [[ObjectRecord]]
-    /// The Object Environment Record is inlined here.
-    pub object_record: Object,
-
-    /// [[GlobalThisValue]]
-    pub global_this_value: Object,
+    /// ### \[\[ObjectRecord\]\]
+    ///
+    /// Binding object is the global object. It contains global built-in
+    /// bindings as well as FunctionDeclaration, GeneratorDeclaration,
+    /// AsyncFunctionDeclaration, AsyncGeneratorDeclaration, and
+    /// VariableDeclaration bindings in global code for the associated realm.
+    object_record: ObjectEnvironment,
 
-    /// [[DeclarativeRecord]]
-    /// The Declaration Environment Record is inlined here.
-    declarative_record: HashMap<String, Binding>,
+    /// ### \[\[GlobalThisValue\]\]
+    ///
+    /// The value returned by this in global scope. Hosts may provide any
+    /// ECMAScript Object value.
+    pub(crate) global_this_value: Object,
 
-    /// [[VarNames]]
-    var_names: ElementsVector,
+    /// ### \[\[DeclarativeRecord\]\]
+    ///
+    /// Contains bindings for all declarations in global code for the associated
+    /// realm code except for FunctionDeclaration, GeneratorDeclaration,
+    /// AsyncFunctionDeclaration, AsyncGeneratorDeclaration, and
+    /// VariableDeclaration bindings.
+    declarative_record: DeclarativeEnvironment,
 
-    /// [[OuterEnv]]
+    /// ### \[\[VarNames\]\]
     ///
-    /// Per https://tc39.es/ecma262/#sec-the-environment-record-type-hierarchy:
```

Comment:
**nitpick:** Could maybe move the spec quote about outer env into the struct description field. The null field doesn't hurt the size of the struct though so it's also an option to keep it here, just for documentation's sake.